drivers/counter/counter_neorv32_gptmr.c

```diff
@@ -0,0 +1,278 @@
+
+#include <soc.h>
+
+LOG_MODULE_REGISTER(neorv32_gptmr, CONFIG_COUNTER_LOG_LEVEL);
+
+/* Registers */
+#define NEORV32_GPTMR_CTRL         0x00
+#define NEORV32_GPTMR_CTRL_EN      BIT(0)
+#define NEORV32_GPTMR_CTRL_PRSC    GENMASK(3, 1)
+#define NEORV32_GPTMR_CTRL_IRQ_CLR BIT(30)
+#define NEORV32_GPTMR_CTRL_IRQ_PND BIT(31)
+
+#define NEORV32_GPTMR_THRES 0x04
+#define NEORV32_GPTMR_COUNT 0x08
+
+struct neorv32_gptmr_config {
+	struct counter_config_info info;
+	const struct device *syscon;
+	mm_reg_t base;
+	uint8_t prescaler;
+	void (*irq_config_func)(void);
+};
+
+struct neorv32_gptmr_data {
+	struct k_spinlock lock;
+	counter_top_callback_t top_callback;
+	void *top_user_data;
+};
+
+static inline uint32_t neorv32_gptmr_read(const struct device *dev, uint16_t reg)
+{
+	const struct neorv32_gptmr_config *config = dev->config;
+
+	return sys_read32(config->base + reg);
+}
+
+static inline void neorv32_gptmr_write(const struct device *dev, uint16_t reg, uint32_t val)
+{
+	const struct neorv32_gptmr_config *config = dev->config;
+
+	sys_write32(val, config->base + reg);
+}
+
+static int neorv32_gptmr_start(const struct device *dev)
+{
+	struct neorv32_gptmr_data *data = dev->data;
+	k_spinlock_key_t key;
+	uint32_t ctrl;
+
+	key = k_spin_lock(&data->lock);
+
+	ctrl = neorv32_gptmr_read(dev, NEORV32_GPTMR_CTRL);
+	ctrl |= NEORV32_GPTMR_CTRL_EN;
+	neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl);
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int neorv32_gptmr_stop(const struct device *dev)
+{
+	struct neorv32_gptmr_data *data = dev->data;
+	k_spinlock_key_t key;
+	uint32_t ctrl;
+
+	key = k_spin_lock(&data->lock);
+
+	ctrl = neorv32_gptmr_read(dev, NEORV32_GPTMR_CTRL);
+	ctrl &= ~NEORV32_GPTMR_CTRL_EN;
+	neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl);
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int neorv32_gptmr_get_value(const struct device *dev, uint32_t *ticks)
+{
+	*ticks = neorv32_gptmr_read(dev, NEORV32_GPTMR_COUNT);
+
+	return 0;
+}
+
+static int neorv32_gptmr_set_top_value(const struct device *dev, const struct counter_top_cfg *cfg)
+{
+	static const uint32_t supported =
+		COUNTER_TOP_CFG_DONT_RESET | COUNTER_TOP_CFG_RESET_WHEN_LATE;
+	struct neorv32_gptmr_data *data = dev->data;
+	k_spinlock_key_t key;
+	bool restart = false;
+	uint32_t count;
+	uint32_t ctrl;
+	int err = 0;
+
+	__ASSERT_NO_MSG(cfg != NULL);
+
+	if (cfg->ticks == 0) {
+		return -EINVAL;
+	}
+
+	if ((cfg->flags & ~supported) != 0U) {
+		LOG_ERR("unsupported flags 0x%08x", cfg->flags);
+		return -ENOTSUP;
+	}
+
+	data->top_callback = cfg->callback;
+	data->top_user_data = cfg->user_data;
+
+	key = k_spin_lock(&data->lock);
+
+	ctrl = neorv32_gptmr_read(dev, NEORV32_GPTMR_CTRL);
+	count = neorv32_gptmr_read(dev, NEORV32_GPTMR_COUNT);
+
+	if ((ctrl & NEORV32_GPTMR_CTRL_EN) != 0U) {
+		if ((cfg->flags & COUNTER_TOP_CFG_DONT_RESET) == 0U) {
+			neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl & ~NEORV32_GPTMR_CTRL_EN);
```

Comment:
There's two branches with:

neorv32_gptmr_write(dev, NEORV32_GPTMR_CTRL, ctrl & ~NEORV32_GPTMR_CTRL_EN);
restart = true;

If it makes sense, move the disable inside the if (restart) block later on.

Reply:
The stop needs to happen before setting the new threshold in this case in order not to miss a top rollover event.